assignment-3/src/data/ArrayListStudentDao.java

```diff
@@ -0,0 +1,78 @@
+package data;
+
+import logic.Student;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Comparator;
+
+public class ArrayListStudentDao implements data.StudentDao {
+    private final List<Student> students = new ArrayList<>();
+
+    @Override
+    public void add(Student student) throws Exception {
+        for (Student s : students) {
+            if (s.getId() == student.getId()) {
+                throw new Exception("Student with ID " + student.getId() + " already exists.");
```

Comment:
Here you should use and throw a custom Exception (e.g. `StudentAlreadyExistsException`).